source/extensions/transport_sockets/tls/cert_validator/default_validator.cc

```diff
@@ -530,7 +532,16 @@ void DefaultCertValidator::addClientValidationContext(SSL_CTX* ctx, bool require
   }
   // Set the verify_depth
   if (config_->maxVerifyDepth().has_value()) {
-    SSL_CTX_set_verify_depth(ctx, config_->maxVerifyDepth().value());
+    int max_verify_depth = std::min(config_->maxVerifyDepth().value(), uint32_t{INT_MAX});
+#if BORINGSSL_API_VERSION >= 29
+    // Older BoringSSLs behave like OpenSSL 1.0.x and exclude the leaf from the
+    // depth but include the trust anchor. Newer BoringSSLs match OpenSSL 1.1.x
+    // and later in excluding both the leaf and trust anchor. `maxVerifyDepth`
+    // documents the older behavior, so adjust the value to match.
+    SSL_CTX_set_verify_depth(ctx, max_verify_depth - 1);
```

Comment:
if `max_verify_depth` is 0, won't this overflow?

Reply:
No, that's why I'm setting it to `int`, contrary to the suggestion in. https://github.com/envoyproxy/envoy/pull/31287#discussion_r1424411253

0 - 1 is perfectly well-defined in C, and the API, at least in BoringSSL's implementation, does work with -1. It's just a bit odd, since it means we'll only accept explicitly-trusted self-signed roots (a questionably well-defined concept in X.509). But, if that's what Envoy is requesting, okay whatever. :)

Reply:
I've added a test for it, so we can keep it vaguely working but as this is a pretty incoherent thing to want, I don't think we'd ever officially support this in BoringSSL, no.

Reply:
If you prefer, I can clamp it to 0 instead of -1. It would mean the new BoringSSL will change your behavior ever-so-slightly, but considering that maxVerifyDepth = 0, under your existing semantics, is incoherent, I don't think the behavior change matters much.

Reply:
I see. Thanks for providing the additional context.
I'll let @ggreenway to weigh in as a senior-maintainer in this domain.

Reply:
I think I'd rather clamp at 0, and take the slight behavior change.

Reply:
Done.